controller/webhook/launcher.go

```diff
@@ -0,0 +1,64 @@
+package webhook
+
+import (
+	"flag"
+	"os"
+	"os/signal"
+	"strconv"
+
+	"github.com/linkerd/linkerd2/controller/k8s"
+	"github.com/linkerd/linkerd2/pkg/admin"
+	"github.com/linkerd/linkerd2/pkg/flags"
+	"github.com/linkerd/linkerd2/pkg/tls"
+	log "github.com/sirupsen/logrus"
+)
+
+// Launch sets up and starts the webhook and metrics servers
+func Launch(config *Config) {
+	p := strconv.FormatUint(uint64(config.MetricsPort), 10)
+	metricsAddr := flag.String("metrics-addr", ":"+p, "address to serve scrapable metrics on")
+	addr := flag.String("addr", ":8443", "address to serve on")
+	kubeconfig := flag.String("kubeconfig", "", "path to kubeconfig")
+	controllerNamespace := flag.String("controller-namespace", "linkerd", "namespace in which Linkerd is installed")
+	flags.ConfigureAndParse()
+
+	stop := make(chan os.Signal, 1)
+	defer close(stop)
+	signal.Notify(stop, os.Interrupt, os.Kill)
+
+	k8sAPI, err := k8s.InitializeAPI(*kubeconfig, k8s.NS, k8s.RS)
+	if err != nil {
+		log.Fatalf("failed to initialize Kubernetes API: %s", err)
+	}
+
+	rootCA, err := tls.GenerateRootCAWithDefaults(config.WebhookServiceName)
+	if err != nil {
+		log.Fatalf("failed to create root CA: %s", err)
+	}
+
+	config.api = k8sAPI
+	config.controllerNamespace = *controllerNamespace
+	config.rootCA = rootCA
+
+	selfLink, err := config.Create()
+	if err != nil {
+		log.Fatalf("failed to create the webhook configurations resource: %s", err)
+	}
+	log.Infof("created webhook configuration: %s", selfLink)
+
+	s, err := NewServer(k8sAPI, *addr, config.WebhookServiceName, *controllerNamespace, rootCA, config.Handler)
+	if err != nil {
+		log.Fatalf("failed to initialize the webhook server: %s", err)
+	}
+
+	k8sAPI.Sync()
+
+	go s.Start()
+	go admin.StartServer(*metricsAddr)
+
+	<-stop
+	log.Info("shutting down webhook server")
+	if err := s.Shutdown(); err != nil {
```

Comment:
we're not very consistent about shutdown across our components, but i think we should move towards passing a context with a timeout into shutdown, we have this pattern today in the `web` component. this at least gives a bit more feedback in case an error is returned (and logged)
https://github.com/linkerd/linkerd2/blob/50952c813e6df45daa26bef43aee6e9f40704ec8/web/main.go#L54-L57